File: Python/Leetcode Medium/11ContainerWithMostWater.py
#https://leetcode.com/problems/container-with-most-water/
class Solution:
    def maxArea(self, height: List[int]) -> int:
        left, right = 0, len(height)-1
        area = min(height[left], height[right]) * (right-left)
        while left < right:
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
            tempArea = min(height[left], height[right]) * (right-left)
            area = max(area, tempArea)
        return area


class Solution:
    def maxArea(self, height: List[int]) -> int:
        area = 0
        for left in range(len(height)):
            for right in range(len(height)):
                tempArea = min(height[left], height[right]) * (right-left)
                area = max(area, tempArea)
        return area 

File: C++/Leetcode Medium/11ContainerWithMostWater.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size()-1;

        int res = 0;
        while (l < r) {
            int leftHeight = height[l];
            int rightHeight = height[r];
            int currVolume = min(leftHeight, rightHeight) * (r-l);
            res = max(res, currVolume);

            if (leftHeight < rightHeight) ++l;
            else --r;
        }

        return res;
    }
};

#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size()-1;
        int res = 0;
        while (l < r) {
            res = max(res, (r-l)*min(height[l], height[r]));
            if (height[l] > height[r]) --r;
            else ++l;
        }
        return res;
    }
};
